feat: retrieve layer 1 gas fee from transaction controller

OGPoyraz opened the pull request:
## **Description**

<!--
Write a short description of the changes included in this pull request, also include relevant motivation and context. Have in mind the following questions:
1. What is the reason for the change?
2. What is the improvement/solution?
-->

This PR is aiming to implement `OptimismLayer1GasFeeFlow` as a first use of `transactionMeta.layer1GasFee`. Hence it contains removal of `fetchEstimatedL1Fee` function and it's helper functions. 

New `transaction-controller` version also includes a `getLayer1GasFee` function which could be used to obtain `layer1GasFeeValue` for un-existing transactions. This is being used in swaps controller. 

Optimism Layer 1 Gas Fee Flow PR in core: https://github.com/MetaMask/core/pull/4055

[![Open in GitHub Codespaces](https://github.com/codespaces/badge.svg)](https://codespaces.new/MetaMask/metamask-extension/pull/23521?quickstart=1)

## **Related issues**

Fixes: https://github.com/MetaMask/MetaMask-planning/issues/1945

## **Manual testing steps**

This PR should not affect existing gas component that shows `Layer1GasFee` amount which appear on send flow, confirmation page & swaps. Hence manually testing any transaction on `Optimism` will work fine. 

## **Screenshots/Recordings**

<!-- If applicable, add screenshots and/or recordings to visualize the before and after of your change. -->

### **Before**

N/A

### **After**

N/A

## **Pre-merge author checklist**

- [X] I’ve followed [MetaMask Coding Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [X] I've clearly explained what problem this PR is solving and how it is solved.
- [X] I've linked related issues
- [X] I've included manual testing steps
- [] I've included screenshots/recordings if applicable
- [X] I’ve included tests if applicable
- [X] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [X] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.
- [X] I’ve properly set the pull request status:
  - [X] In case it's not yet "ready for review", I've set it to "draft".
  - [X] In case it's "ready for review", I've changed it from "draft" to "non-draft".

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

OGPoyraz committed:
Use transaction meta layer1GasFee

OGPoyraz committed:
Remove recently added fetch method

OGPoyraz committed:
remove old changes

OGPoyraz committed:
revert controller

OGPoyraz committed:
latest changes

OGPoyraz committed:
update yarn.lock

OGPoyraz committed:
yarn dedupe

OGPoyraz committed:
Update MMcontroller

OGPoyraz committed:
update swaps tests

OGPoyraz committed:
update swaps tests

OGPoyraz committed:
fix lint

OGPoyraz committed:
fix swaps tests

OGPoyraz committed:
fix swaps tests

OGPoyraz committed:
fix fee detail component

OGPoyraz commented:
@metamaskbot update-policies

sleepytanya commented:
Dev build [git id]([1591ae9](https://github.com/MetaMask/metamask-extension/pull/23521/commits/1591ae9fac26fa53fd97b80a462d0b3eceb5aba3)), Chrome, Firefox - Optimism Layer1 Gas fees displayed in the Send flow:
<img width="372" alt="Screenshot 2024-03-19 at 12 56 53 PM" src="https://github.com/MetaMask/metamask-extension/assets/104780023/3d0242c3-007d-4e35-b47e-9ca59fe13497">
<img width="445" alt="Screenshot 2024-03-19 at 12 55 51 PM" src="https://github.com/MetaMask/metamask-extension/assets/104780023/f8f3efbb-3cc6-4db1-a915-abf4ce30b632">

OGPoyraz committed:
Update package & policies

OGPoyraz commented:
@metamaskbot update-policies

OGPoyraz committed:
Merge branch 'develop' into use-layer1gasfeeflow

OGPoyraz commented:
@metamaskbot update-policies

OGPoyraz committed:
update yarn lock

OGPoyraz committed:
Merge branch 'develop' into use-layer1gasfeeflow

OGPoyraz commented:
@metamaskbot update-policies

OGPoyraz committed:
update fee-details component

OGPoyraz committed:
Merge branch 'use-layer1gasfeeflow' of github.com:MetaMask/metamask-extension into use-layer1gasfeeflow

metamaskbot committed:
Update LavaMoat policies

OGPoyraz committed:
update lock file

OGPoyraz committed:
update privacy snapshot

OGPoyraz committed:
revert privacy snapshot change

OGPoyraz committed:
Merge branch 'develop' into use-layer1gasfeeflow

OGPoyraz committed:
update lock

OGPoyraz committed:
Fix lint

OGPoyraz committed:
Remove unused dep

OGPoyraz committed:
Merge branch 'develop' into use-layer1gasfeeflow

OGPoyraz commented:
@metamaskbot update-policies

metamaskbot committed:
Update LavaMoat policies

matthewwalsh0 commented on app/scripts/controllers/swaps.js:
```diff
@@ -256,6 +260,7 @@ export default class SwapsController {
     isPolledRequest,
   ) {
     const { chainId } = fetchParamsMetaData;
+    const networkClientId = this._getNetworkClientId();
```
Apologies, I think I missed this new method in the core PR.

Ideally `networkClientId` is an optional argument, meaning the clients that aren't fully multi-chain can instead just pass the `chainId` and we use just that to find a suitable network client in the background.

Meaning this additional constructor option wouldn't be needed.

matthewwalsh0 replied:
There was a small error that needed fixing in the `TransactionController`, so I've updated the signature to make these optional at the same time.

matthewwalsh0 commented on app/scripts/metamask-controller.js:
```diff
@@ -3191,6 +3196,7 @@ export default class MetamaskController extends EventEmitter {
         txController.updatePreviousGasParams.bind(txController),
       abortTransactionSigning:
         txController.abortTransactionSigning.bind(txController),
+      getLayer1GasFee: txController.getLayer1GasFee.bind(txController),
```
Since we have so many properties here, we'll have to go alphabetical eventually... 😄

matthewwalsh0 commented on ui/ducks/send/send.js:
```diff
@@ -512,11 +513,12 @@ export const computeEstimatedGasLimit = createAsyncThunk(
     const isNonStandardEthChain = getIsNonStandardEthChain(state);
     const chainId = getCurrentChainId(state);
     const selectedAccount = getSelectedInternalAccountWithBalance(state);
+    const networkClientId = getSelectedNetworkClientId(state);
```
As mentioned before, this argument isn't adding any multi-chain support yet since it just references the global network client ID.

So ideally it would be optional and only the `chainId` would be needed.

matthewwalsh0 replied:
Done.

matthewwalsh0 commented on ui/pages/confirmations/components/fee-details-component/fee-details-component.js:
```diff
@@ -27,42 +27,30 @@ import LoadingHeartBeat from '../../../../components/ui/loading-heartbeat';
 }) {
+  const layer1GasFee = txData?.layer1GasFee;
   const [expandFeeDetails, setExpandFeeDetails] = useState(false);
-  const [estimatedL1Fees, setEstimatedL1Fees] = useState(null);
```
I assume our empty value is now `undefined` if not set where as it was `null` before.

Should we use a `?? null` above for backwards compatibility?

matthewwalsh0 replied:
Done.

matthewwalsh0 commented on ui/pages/confirmations/components/fee-details-component/fee-details-component.js:
```diff
@@ -148,13 +136,13 @@ export default function FeeDetailsComponent({
+                useCurrencyRateCheck && renderTotalDetailText(layer1GasFee)
               }
-              detailTotal={renderTotalDetailValue(estimatedL1Fees)}
+              detailTotal={renderTotalDetailValue(layer1GasFee)}
```
Just for audit sake, I assume the previous value and our value are both in WEI and hexadecimal?

matthewwalsh0 commented on ui/pages/confirmations/components/multilayer-fee-message/multi-layer-fee-message.js:
```diff
@@ -19,50 +17,27 @@ export default function MultilayerFeeMessage({
-        );
-        setLayer1Total(result);
-      } catch (e) {
-        captureException(e);
```
Unfortunately we do lose any specific error tracking if the flow fails, but we could add suitable properties to the metadata if it becomes an issue.

For example, we could have an `layer1Fee: {value?: Hex, error?: string}`.

matthewwalsh0 commented on ui/pages/confirmations/components/multilayer-fee-message/multi-layer-fee-message.test.js:
```diff
@@ -5,11 +5,6 @@ import { renderWithProvider } from '../../../../../test/lib/render-helpers';
 
-jest.mock(
-  '../../../../helpers/utils/optimism/fetchEstimatedL1Fee',
-  () => '0x5',
```
Do we need to replace this by adding the `layer1Fee` property to a mock transaction somewhere?

matthewwalsh0 replied:
Done.

matthewwalsh0 commented on ui/pages/confirmations/confirm-transaction-base/confirm-transaction-base.component.js:
```diff
@@ -259,21 +256,11 @@ export default class ConfirmTransactionBase extends Component {
+      hexMaximumTransactionFee !== prevHexMaximumTransactionFee &&
+      useMaxValue
+    ) {
+      this.updateValueToMax();
```
Do we need to update the send max logic to also subtract the new L1 fee property or is that already done?

matthewwalsh0 replied:
This may be a pre-existing bug we can address in a different PR.

matthewwalsh0 commented on ui/store/actions.ts:
```diff
@@ -2203,6 +2203,21 @@ export function abortTransactionSigning(
+  chainId: Hex,
+  networkClientId: NetworkClientId,
+  transactionParams: TransactionParams,
+  // TODO: Replace `any` with type
```
Is this needed, has `any` been disallowed in the extension also?

matthewwalsh0 commented on app/scripts/metamask-controller.js:
```diff
@@ -3191,6 +3196,7 @@ export default class MetamaskController extends EventEmitter {
         txController.updatePreviousGasParams.bind(txController),
       abortTransactionSigning:
         txController.abortTransactionSigning.bind(txController),
+      getLayer1GasFee: txController.getLayer1GasFee.bind(txController),
```
Concerning the comments about `networkClientId` ideally being optional.

Pending a new controller release, could we instead make it optional here by populating it within an arrow function if not provided?

We could read it directly from the `NetworkController` state as we currently do above?

matthewwalsh0 replied:
Not required as the controller signature has been updated.

matthewwalsh0 commented on ui/pages/confirmations/components/fee-details-component/fee-details-component.js:
I assume there is no storybook stories that were relying on the old implementation that we have to update?

Otherwise we'd see failures in the storybook job.

matthewwalsh0 committed:
Merge branch 'develop' into use-layer1gasfeeflow

matthewwalsh0 committed:
Merge branch 'develop' into use-layer1gasfeeflow

matthewwalsh0 committed:
Add patch to fix layer 1 update from polling

Update calls to getLayer1GasFee to use request object.
Remove unnecessary simulation logic.

matthewwalsh0 committed:
Fix linting

matthewwalsh0 commented:
@metamaskbot update-policies

metamaskbot committed:
Update LavaMoat policies

matthewwalsh0 committed:
Fix linting

matthewwalsh0 committed:
Merge branch 'develop' into use-layer1gasfeeflow

cryptotavares commented:
why do we still need the `getIsMultiLayerFeeNetwork` selector? Couldn't we just tell if the `layer1GasFee` is not undefined? 🤔

matthewwalsh0 commented:
> why do we still need the `getIsMultiLayerFeeNetwork` selector? Couldn't we just tell if the `layer1GasFee` is not undefined? 🤔

That's already ready in an additional PR as it requires alot of changes also and test updates etc.

sleepytanya commented:
Latest build - Optimism Layer1 fees displayed on confirmations screen:

<img width="633" alt="Screenshot 2024-04-12 at 12 11 51 PM" src="https://github.com/MetaMask/metamask-extension/assets/104780023/a7709730-c201-4703-b3ec-e9ffdf08cb52">

<img width="632" alt="Screenshot 2024-04-12 at 12 22 49 PM" src="https://github.com/MetaMask/metamask-extension/assets/104780023/4167d91d-cc0a-4682-a5c7-0c1832175df7">

matthewwalsh0 committed:
Use new controller release

matthewwalsh0 committed:
Remove yarn duplication

matthewwalsh0 commented:
@metamaskbot update-policies

metamaskbot committed:
Update LavaMoat policies
